flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsOrRecordSerializer.java

```diff
@@ -0,0 +1,219 @@
+import java.util.Objects;
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.typeutils.CompositeTypeSerializerSnapshot;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.api.common.typeutils.TypeSerializerSnapshot;
+import org.apache.flink.core.memory.DataInputView;
+import org.apache.flink.core.memory.DataOutputView;
+import org.apache.flink.table.data.RowData;
+
+@Internal
+class DataStatisticsOrRecordSerializer<D extends DataStatistics<D, S>, S>
+    extends TypeSerializer<DataStatisticsOrRecord<D, S>> {
+  private final TypeSerializer<DataStatistics<D, S>> statisticsSerializer;
+  private final TypeSerializer<RowData> recordSerializer;
+
+  DataStatisticsOrRecordSerializer(
+      TypeSerializer<DataStatistics<D, S>> statisticsSerializer,
+      TypeSerializer<RowData> recordSerializer) {
+    this.statisticsSerializer = statisticsSerializer;
+    this.recordSerializer = recordSerializer;
+  }
+
+  @Override
+  public boolean isImmutableType() {
+    return false;
+  }
+
+  @SuppressWarnings("ReferenceEquality")
+  @Override
+  public TypeSerializer<DataStatisticsOrRecord<D, S>> duplicate() {
+    TypeSerializer<DataStatistics<D, S>> duplicateStatisticsSerializer =
+        statisticsSerializer.duplicate();
+    TypeSerializer<RowData> duplicateRowDataSerializer = recordSerializer.duplicate();
+    if ((statisticsSerializer != duplicateStatisticsSerializer)
+        || (recordSerializer != duplicateRowDataSerializer)) {
+      return new DataStatisticsOrRecordSerializer<>(
+          duplicateStatisticsSerializer, duplicateRowDataSerializer);
+    } else {
+      return this;
+    }
+  }
+
+  @Override
+  public DataStatisticsOrRecord<D, S> createInstance() {
+    // arbitrarily always create RowData value instance
+    return DataStatisticsOrRecord.fromRecord(recordSerializer.createInstance());
+  }
+
+  @Override
+  public DataStatisticsOrRecord<D, S> copy(DataStatisticsOrRecord<D, S> from) {
+    if (from.hasRecord()) {
+      return DataStatisticsOrRecord.fromRecord(recordSerializer.copy(from.record()));
+    } else {
+      return DataStatisticsOrRecord.fromDataStatistics(
+          statisticsSerializer.copy(from.dataStatistics()));
+    }
+  }
+
+  @Override
+  public DataStatisticsOrRecord<D, S> copy(
+      DataStatisticsOrRecord<D, S> from, DataStatisticsOrRecord<D, S> reuse) {
+    DataStatisticsOrRecord<D, S> to;
+    if (from.hasRecord()) {
+      to = DataStatisticsOrRecord.reuseRecord(reuse, recordSerializer);
+      RowData record = recordSerializer.copy(from.record(), to.record());
+      to.record(record);
+    } else {
+      to = DataStatisticsOrRecord.reuseStatistics(reuse, statisticsSerializer);
+      DataStatistics<D, S> statistics =
+          statisticsSerializer.copy(from.dataStatistics(), to.dataStatistics());
+      to.dataStatistics(statistics);
+    }
+
+    return to;
+  }
+
+  @Override
+  public int getLength() {
+    return -1;
```

Comment:
We only care about the amount of data, not the size of data, right?

Reply:
@hililiwei not sure if I understand the question here. can you elaborate?

here is the javadoc for this method from the TypeSerializer interface.
```
Returns:
The length of the data type, or -1 for variable length data types.
```

Reply:
@hililiwei not sure if I understand the question here. can you elaborate?

here is the javadoc for this method from the Flink `TypeSerializer` interface.
```
Returns:
The length of the data type, or -1 for variable length data types.
```

Reply:
@hililiwei not sure if I understand the question here. can you elaborate?

here is the javadoc for this method from the Flink `TypeSerializer` interface.
```
Returns:
The length of the data type, or -1 for variable length data types.
```